modules/table/src/test/java/org/apache/ignite/internal/table/distributed/AbstractMvStorageUpdateHandlerTest.java

```diff
@@ -0,0 +1,144 @@
+            @InjectConfiguration DistributionZoneConfiguration distributionZoneConfig
+    ) {
+        storageEngine = createStorageEngine();
+
+        storageEngine.start();
+
+        TableConfiguration tableConfig = tablesConfig.tables().get("foo");
+
+        assertThat(
+                tableConfig.dataStorage().change(dataStorageChange -> dataStorageChange.convert(storageEngine.name())),
+                willCompleteSuccessfully()
+        );
+
+        MvTableStorage mvTableStorage = storageEngine.createMvTable(tableConfig, tablesConfig, distributionZoneConfig);
+
+        mvTableStorage.start();
+
+        MvPartitionStorage mvPartitionStorage = getOrCreateMvPartition(mvTableStorage, PARTITION_ID);
+
+        partitionDataStorage = new TestPartitionDataStorage(mvPartitionStorage);
+
+        storageUpdateHandler = new StorageUpdateHandler(PARTITION_ID, partitionDataStorage, Map::of, tableConfig.dataStorage());
+    }
+
+    @AfterEach
+    void tearDown() {
+        if (storageEngine != null) {
+            storageEngine.stop();
+        }
+    }
+
+    protected abstract StorageEngine createStorageEngine();
+
+    @Test
+    void testConcurrentVacuumBatch() {
+        RowId rowId0 = new RowId(PARTITION_ID);
+        RowId rowId1 = new RowId(PARTITION_ID);
+
+        BinaryRow row0 = binaryRow(new TestKey(0, "key0"), new TestValue(0, "value0"));
+        BinaryRow row1 = binaryRow(new TestKey(0, "key0"), new TestValue(0, "value0"));
+
+        for (int i = 0; i < REPEATS; i++) {
+            addWriteCommitted(partitionDataStorage, rowId0, row0, clock.now());
+            addWriteCommitted(partitionDataStorage, rowId1, row1, clock.now());
+
+            addWriteCommitted(partitionDataStorage, rowId0, row0, clock.now());
+            addWriteCommitted(partitionDataStorage, rowId1, row1, clock.now());
+
+            addWriteCommitted(partitionDataStorage, rowId0, row0, clock.now());
+            addWriteCommitted(partitionDataStorage, rowId1, row1, clock.now());
+
+            addWriteCommitted(partitionDataStorage, rowId0, row0, clock.now());
+            addWriteCommitted(partitionDataStorage, rowId1, row1, clock.now());
+
+            addWriteCommitted(partitionDataStorage, rowId0, row0, clock.now());
+            addWriteCommitted(partitionDataStorage, rowId1, row1, clock.now());
+
+            addWriteCommitted(partitionDataStorage, rowId0, null, clock.now());
+            addWriteCommitted(partitionDataStorage, rowId1, null, clock.now());
+
+            runRace(
+                    () -> storageUpdateHandler.vacuumBatch(HybridTimestamp.MAX_VALUE, 2),
+                    () -> storageUpdateHandler.vacuumBatch(HybridTimestamp.MAX_VALUE, 2),
+                    () -> storageUpdateHandler.vacuumBatch(HybridTimestamp.MAX_VALUE, 2),
+                    () -> storageUpdateHandler.vacuumBatch(HybridTimestamp.MAX_VALUE, 2),
+                    () -> storageUpdateHandler.vacuumBatch(HybridTimestamp.MAX_VALUE, 2)
```

Comment:
Having 5 threads is unrealistic, we only have two in real system

Reply:
Fix it